File: LeetCode/Concurrency/1226. The Dining Philosophers.cpp
// 1226. The Dining Philosophers
class DiningPhilosophers {
public:

    int n;
    vector<std::mutex> mtx;

    DiningPhilosophers() {
        n = 5;
        mtx = vector<std::mutex> (n);
    }

    void wantsToEat(int philosopher,
                    function<void()> pickLeftFork,
                    function<void()> pickRightFork,
                    function<void()> eat,
                    function<void()> putLeftFork,
                    function<void()> putRightFork) {

        int leftFork = philosopher%n;
        int rightFork = (philosopher + 1)%n;

        //for Even philosopher's first pick left then right and keep in that order
        if(philosopher%2==0){
            mtx[leftFork].lock();
            mtx[rightFork].lock();

            pickLeftFork();
            pickRightFork();
            eat();
            putLeftFork();
            putRightFork();

            mtx[leftFork].unlock();
            mtx[rightFork].unlock();
        }
        //for Odd philosopher's first pick right then left and keep in that order
        else {
            mtx[rightFork].lock();
            mtx[leftFork].lock();

            pickRightFork();
            pickLeftFork();
            eat();
            putRightFork();
            putLeftFork();

            mtx[rightFork].unlock();
            mtx[leftFork].unlock();
        }
    }
};

File: LeetCode/Concurrency/1226. The Dining Philosophers.java
// 1226. The Dining Philosophers

class DiningPhilosophers {

    private int n;
    private ReentrantLock[] rntLck;
    private Semaphore eating;

    public DiningPhilosophers() {
        n = 5;
        eating = new Semaphore (n-1);
        rntLck = new ReentrantLock[n];
        for(int i = 0; i < n; i++) {
            rntLck[i] = new ReentrantLock();
        }   
    }

    // call the run() method of any runnable to execute its code
    public void wantsToEat(int philosopher,
                           Runnable pickLeftFork,
                           Runnable pickRightFork,
                           Runnable eat,
                           Runnable putLeftFork,
                           Runnable putRightFork) throws InterruptedException {

        int leftFork = philosopher % n;
        int rightFork = (philosopher + 1) % n;
        
        eating.acquire();

        rntLck[leftFork].lock();
        rntLck[rightFork].lock();

        pickLeftFork.run();
        pickRightFork.run();
        eat.run();
        putLeftFork.run();
        putRightFork.run();

        rntLck[leftFork].unlock();
        rntLck[rightFork].unlock();
       
        eating.release();
    }
}